node/src/components/consensus/highway_core/active_validator.rs

```diff
@@ -478,82 +483,268 @@ mod tests {
+            let proposal_timestamp = block_context.timestamp();
+            let effects = validator.propose(
+                cv,
+                block_context,
+                &self.state,
+                self.instance_id,
+                &mut self.rng,
+            );
+
+            // Add the new unit to the state.
+            let proposal_wunit = unwrap_single(&effects).unwrap_unit();
+            let prop_hash = proposal_wunit.hash();
+            self.state.add_unit(proposal_wunit.clone()).unwrap();
+            let effects = validator.on_new_unit(
+                &prop_hash,
+                proposal_timestamp + 1.into(),
+                &self.state,
+                self.instance_id,
+                &mut self.rng,
+            );
+            self.schedule_timer(vidx, &effects);
+            (effects, proposal_wunit)
+        }
+
+        /// Handle new unit by validator `vidx`.
+        /// Since all validator use the same state, that unit should be added already. Panics if
+        /// not. Returns effect created when handling new unit.
+        fn handle_new_unit(
+            &mut self,
+            vidx: ValidatorIndex,
+            uhash: &<TestContext as Context>::Hash,
+        ) -> Vec<Effect<TestContext>> {
+            let validator = self.validators.get_mut(&vidx).unwrap();
+            let delivery_timestamp = self.state.unit(uhash).timestamp + 1.into();
+            let effects = validator.on_new_unit(
+                uhash,
+                delivery_timestamp,
+                &self.state,
+                self.instance_id,
+                &mut self.rng,
+            );
+            self.schedule_timer(vidx, &effects);
+            self.add_new_unit(&effects);
+            effects
+        }
+
+        /// Schedules new timers, if any was returned as an effect.
+        fn schedule_timer(&mut self, vidx: ValidatorIndex, effects: &[Effect<TestContext>]) {
+            let new_timestamps: Vec<Timestamp> = effects
+                .iter()
+                .filter_map(|eff| {
+                    if let Effect::ScheduleTimer(timestamp) = eff {
+                        Some(*timestamp)
+                    } else {
+                        None
+                    }
+                })
+                .collect();
+            match *new_timestamps {
+                [] => (),
+                [timestamp] => {
+                    let _ = self.timers.insert((timestamp, vidx));
+                }
+                _ => panic!(
+                    "Expected at most one timer to be scheduled: {:?}",
+                    new_timestamps
+                ),
+            }
+        }
+
+        /// Adds new unit, if any, to the state.
+        fn add_new_unit(&mut self, effects: &[Effect<TestContext>]) {
+            let new_units: Vec<_> = effects
+                .iter()
+                .filter_map(|eff| {
+                    if let Effect::NewVertex(ValidVertex(Vertex::Unit(swunit))) = eff {
```

Comment:
Should we also add new endorsements?

Reply:
I plan on doing that but didn't want to do it in _this_ PR that is only introducing the wrapper and doesn't change the logic that was already here.